Realm/RLMDictionary.mm

```diff
@@ -0,0 +1,559 @@
+            options:(NSKeyValueObservingOptions)options context:(void *)context {
+    RLMDictionaryValidateObservationKey(keyPath, self);
+    [super addObserver:observer forKeyPath:keyPath options:options context:context];
+}
+
+#pragma mark - Methods unsupported on unmanaged RLMDictionary instances
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wunused-parameter"
+
+- (nonnull RLMResults *)objectsWithPredicate:(nonnull NSPredicate *)predicate {
+    @throw RLMException(@"This method may only be called on RLMDictionary instances retrieved from an RLMRealm");
+}
+
+- (RLMResults *)sortedResultsUsingDescriptors:(nonnull NSArray<RLMSortDescriptor *> *)properties {
+    @throw RLMException(@"This method may only be called on RLMDictionary instances retrieved from an RLMRealm");
+}
+
+- (RLMResults *)sortedResultsUsingKeyPath:(nonnull NSString *)keyPath ascending:(BOOL)ascending {
+    @throw RLMException(@"This method may only be called on RLMDictionary instances retrieved from an RLMRealm");
+}
+
+- (RLMResults *)distinctResultsUsingKeyPaths:(NSArray<NSString *> *)keyPaths {
+    @throw RLMException(@"This method may only be called on RLMDictionary instances retrieved from an RLMRealm");
+}
+
+// The compiler complains about the method's argument type not matching due to
+// it not having the generic type attached, but it doesn't seem to be possible
+// to actually include the generic type
+// http://www.openradar.me/radar?id=6135653276319744
+#pragma clang diagnostic ignored "-Wmismatched-parameter-types"
+- (nonnull RLMNotificationToken *)addNotificationBlock:(nonnull void (^)(RLMDictionary *, RLMCollectionChange *, NSError *))block {
+    @throw RLMException(@"This method may only be called on RLMDictionary instances retrieved from an RLMRealm");
+}
+
+- (RLMNotificationToken *)addNotificationBlock:(void (^)(RLMDictionary *, RLMCollectionChange *, NSError *))block
+                                         queue:(nullable dispatch_queue_t)queue {
+    @throw RLMException(@"This method may only be called on RLMDictionary instances retrieved from an RLMRealm");
+}
+
+- (instancetype)freeze {
+    @throw RLMException(@"This method may only be called on RLMDictionary instances retrieved from an RLMRealm");
+}
+
+- (instancetype)thaw {
+    @throw RLMException(@"This method may only be called on RLMDictionary instances retrieved from an RLMRealm");
+}
+
+- (nonnull id)objectAtIndex:(NSUInteger)index {
+    validateDictionaryBounds(self, index);
+    NSString *key = _backingCollection.allKeys[index];
+    return _backingCollection[key];
+}
+
+- (NSUInteger)indexOfObject:(id)value {
+    return [_backingCollection.allValues indexOfObject:value];
+}
+
+#pragma clang diagnostic pop // unused parameter warning
+
+#pragma mark - Thread Confined Protocol Conformance
+
+- (realm::ThreadSafeReference)makeThreadSafeReference {
+    REALM_TERMINATE("Unexpected handover of unmanaged `RLMDictionary`");
```

Comment:
This is the same message as RLMArray and RLMSet. I think its self explanatory but maybe changing 'Unexpected' to 'Unsupported' might work better?

Reply:
This is REALM_TERMINATE and not an exception because it's not possible to hit this code path unless you've bypassed the public API. There's a (hopefully) better error message that tries to explain what's wrong at the public API boundary.